snake.asm

```diff
@@ -1,42 +1,62 @@
+std                     ; Set the direction flag to reverse the operation of lodsw for FIFO processing
+start:                  ; Start new game setup
+mov ax, 0x3             ; Set ah=0 (change video mode) and al=3 (80x25 16 color text mode)
+int 0x10                ; Initiate Video BIOS interrupt
+mov di, 0x7D0           ; Set the initial position of the snake's head at the middle of the screen (0xFA0/2)
+mov si, sp              ; Setting si to the stack pointer for loading stack data using lodsw
+.food:                  ; Position new food
+imul bx, sp             ; Multiply bx by sp to pseudo-randomize bx, which points to the food position
+and bx, cx              ; Make bx divisible by 4 to fit on the screen (align with character positions). After anding, we now know that bx < cx (size of the screen) therefore it fits inside
+cmp [bx], ch            ; Compare character at memory location indicated by bx with snake body character
+; Check if the calculated position (bx) is already occupied by the snake body
+; If true generate a new random position for placing food by jumping to the .food label
+je .food                ; If the comparison is true, jump to .food label to generate new food position
+mov [bx], cl            ; Place a food character at the calculated position
+; Now that we've ensured the calculated position (bx) is valid and not occupied,
+; we can place a food character at that location on the display to indicate the new position for the snake's food
+.input:                 ; Process keyboard input
+in al, 0x60             ; Read keyboard input from port 0x60
+mov bx, 0x4             ; Set default distance for snake movement to 4 (leftwards)
+and al, 0x1E            ; Mask out non-arrow key bits (up, down, left, right)
+jp $+0x4                ; Jump if the parity flag is set (skip the next instruction if input is an arrow key)
+mov bl, cl              ; Set distance for horizontal movement based on screen width (cl)
+and al, 0x14            ; Extract only the left and right arrow key bits
+jz $+0x4                ; Jump if the zero flag is set (skip the next instruction if input is left or right)
+neg bx                  ; Invert direction for left or right movement
+sub di, bx              ; Update snake's head position based on movement
+cmp di, cx              ; Check if snake's head position is beyond screen boundaries
+ja start                ; If out of bounds, restart the game
+; The comparison is unsigned (ja) to determine if snake's head has moved beyond screen boundaries
+; In unsigned comparisons, only the magnitude of the values is considered, rather their signs
+; this is crucial, because memory addresses like di are treated as unsigned, so the comparison
+; ensures that if di > cx (screen size) or di < 0, the jump will happen, preventing the snake from
+; moving outside the screen area.

```

Comment:
I do not think it's inaccurate but just doesn't delve more in unsigned, it can be described basically as "if an unsigned value is negative it will wrap-around (2 ^ n - absolute value of the negative number (without the sign)) to prevent it from overflowing the max range of it's bit representation. In our case n is 16 since `di` (destination index) is 16 bit"